include/exadg/acoustic_conservation_equations/postprocessor/sound_energy_calculator.h

```diff
@@ -0,0 +1,263 @@
+        unsigned int const                      block_index_velocity_in)
+  {
+    time_control.setup(data_in.time_control_data);
+
+    matrix_free = &matrix_free_in;
+    data        = data_in;
+
+    dof_index_pressure   = dof_index_pressure_in;
+    dof_index_velocity   = dof_index_velocity_in;
+    quad_index           = quad_index_in;
+    block_index_pressure = block_index_pressure_in;
+    block_index_velocity = block_index_velocity_in;
+
+    rho       = static_cast<Number>(data_in.density);
+    c         = static_cast<Number>(data_in.speed_of_sound);
+    rhocc_inv = static_cast<Number>(
+      1.0 / (data_in.density * data_in.speed_of_sound * data_in.speed_of_sound));
+
+    clear_files = data_in.clear_file;
+
+    if(data_in.time_control_data.is_active)
+    {
+      AssertThrow(data_in.density > 0.0 && data_in.speed_of_sound > 0.0,
+                  dealii::ExcMessage("Material parameters not set in SoundEnergyCalculatorData."));
+
+      create_directories(data_in.directory, mpi_comm);
+    }
+  }
+
+  void
+  evaluate(BlockVectorType const & solution, double const & time, bool const unsteady)
+  {
+    AssertThrow(unsteady,
+                dealii::ExcMessage(
+                  "This postprocessing tool can only be used for unsteady problems."));
+
+    do_evaluate(solution, time);
+  }
+
+  TimeControl time_control;
+
+private:
+  void
+  do_evaluate(BlockVectorType const & solution, double const time)
+  {
+    double sound_energy = calculate_sound_energy(solution);
+
+    // write output file
+    if(dealii::Utilities::MPI::this_mpi_process(mpi_comm) == 0)
+    {
+      std::ostringstream filename;
+      filename << data.directory + data.filename;
+
+      std::ofstream f;
+      if(clear_files == true)
+      {
+        f.open(filename.str().c_str(), std::ios::trunc);
+        f << "Time, Sound energy: E = (1,p*p/(2*rho*c*c)+rho*u*u/2)_Omega" << std::endl;
+        clear_files = false;
+      }
+      else
+      {
+        f.open(filename.str().c_str(), std::ios::app);
+      }
+
+      unsigned int precision = 12;
+      f << std::scientific << std::setprecision(precision) << std::setw(precision + 8) << time
+        << ", " << std::setw(precision + 8) << sound_energy << std::endl;
+    }
+  }
+
+
+  /*
+   *  This function calculates the sound energy
+   *
+   *  Sound energy: E = (1,p*p/(2*rho*c*c)+rho*u*u/2)_Omega
+   *
+   */
+  double
+  calculate_sound_energy(BlockVectorType const & solution) const
+  {
+    Number energy = 0.;
+    matrix_free->cell_loop(&This::cell_loop, this, energy, solution);
+
+    // sum over all MPI processes
+    double const sound_energy = dealii::Utilities::MPI::sum(energy, mpi_comm);
+
+    return sound_energy;
+  }
+
+  void
+  cell_loop(dealii::MatrixFree<dim, Number> const &       matrix_free_in,
+            Number &                                      dst,
+            BlockVectorType const &                       src,
```

Comment:
I believe the implementation would be simpler and more flexible overall when implementing this class by using separate velocity and pressure DoF-vectors.

Reply:
I was thinking about this as well. I decided to do it this way because otherwise we would have to store one of the two DoF vectors as a member (instead of directly passing one `BlockVector` to the loop. On the other hand we have to provide the block indices. 

I think both versions have their pros and cons and can adapt to whatever you prefer.

Reply:
I see a difference between the two variants because as far as I know the transformation between normal DoFVectors and BlockVectors does not work equivalently in both directions from the dealii-side.  So I consider the variant with the two dof vectors more flexible. Or is there functionality to create a block-vector out of two dof-vectors "on-the-fly" without copying the vectors (a constructor like `BlockVector(std::vector<DealiiDofVector *>)`)?

> because otherwise we would have to store one of the two DoF vectors as a member

I think technically there is a solution to this problem. You can pass to `cell_loop()` an argument like `std::vector<VectorType const*> const &  src`, which you can initiliaze prior to calling cell-loop via `std::vector<VectorType const *> src{&velocity, &pressure};`. This way, we should be able to write the interface of this postprocessing routine in a flexible way with `velocity` and `pressure` dof-vectors and without block-indices. Choosing this design, the indexing can be done purely local and would then be a detail of the implementation of  this postprocessing routine.